lib/ansible/inventory/group.py

```diff
@@ -112,7 +112,8 @@ def _check_children_depth(self):
             raise AnsibleError("The group named '%s' has a recursive dependency loop." % self.name)
 
     def add_host(self, host):
-
+        if host in self.hosts:
+            return
```

Comment:
Over in `inventory.data.InventoryData.add_child` we weren't doing any checks to ensure the host wasn't already in the group before adding it.  Instead of spreading the logic elsewhere, this check will ensure the host isn't in the list of hosts in the group before adding again.